service/storage_service.cc

```diff
@@ -752,35 +752,38 @@ future<> storage_service::merge_topology_snapshot(raft_snapshot snp) {
     if (it != snp.mutations.end()) {
         auto s = _db.local().find_schema(db::system_keyspace::NAME, db::system_keyspace::CDC_GENERATIONS_V3);
 
-        // Split big mutations into smaller ones, prepare frozen_muts_to_apply
-        std::vector<frozen_mutation> frozen_muts_to_apply;
+        // Split big mutations into smaller ones
+        // Apply non-atomically so as not to hit the commitlog size limit.
+        // The cdc_generations_v3 data is not used in any way until
+        // it's referenced from the topology table.
+        // By applying the cdc_generations_v3 mutations before topology mutations
+        // we ensure that the lack of atomicity isn't a problem here.
         {
+            semaphore sem(128);
             const auto max_size = _db.local().schema_commitlog()->max_record_size() / 2;
+            future<> done = make_ready_future<>();
+            auto apply_mutation = [&] (frozen_mutation m) -> future<> {
+                auto units = get_units(sem, 1);
+                co_await _db.local().apply(s, m, {}, db::commitlog::force_sync::yes, db::no_timeout);
+            };
+            auto apply_in_background = [&] (frozen_mutation m) {
+                done = done.then([&, m = std::move(m)] () mutable {
```

Comment:
Yeah :(

I also don't understand, how is this better from calling `max_concurrent_for_each` in the foreground? The previous code was simpler. The commit message does not explain how doing this background thing (but then actually waiting for it) is an improvement.

Reply:
Ah, so we basically replace a vector of frozen_mutations, with a list of tasks (`then` continuation chain). What's wrong with the vector though? The vector should be small anyway, even if it's "long" (but it won't ever be that long, really) because mutations are by themselves small objects IIUC (in the sense that they are tree-based data structures, small pointers to larger objects).

The complexity is IMO not worth it, and error prone (e.g. accidentally making everything sequential)

Reply:
The question is how many cdc mutations can we have and how much external memory they hold, not the size of the vector capacity per-se.
I'm concerned about a large number of tables with cdc enabled.

Reply:
> Since we wait for previous "done", there is no parallelism, mutations are applied sequentially.
> This can increase latency of snapshot application. \cc @kbr-scylla

The intention was the call `apply_mutation` and enqueuing only the returned future.
That said, we can consider this as a follow-up.
It's not part of the core changes of this PR.